pyat/at/physics/energy_loss.py

```diff
@@ -123,40 +128,50 @@ def singlev(values):
             raise AtError('values not equal for all cavities')
         return vals[0]
 
-    def eq(x):
-        return numpy.sum(rfv*numpy.sin(2*pi*freq*(x+tl0)/clight))-u0
-
-    def deq(x):
-        return numpy.sum(rfv*numpy.cos(2*pi*freq*(x+tl0)/clight))
+    def eq(x, freq, rfv, tl0, u0):
+        omf = 2*numpy.pi*freq/clight
+        eq1 = numpy.sum(-rfv*numpy.sin(omf*(x-tl0)))-u0
+        eq2 = numpy.sum(-omf*rfv*numpy.cos(omf*(x-tl0)))
+        if eq2 > 0:
+            return numpy.sqrt(eq1**2+eq2**2)
+        else:
+            return eq1
 
     if cavpts is None:
-        cavpts = get_cells(ring, checktype(RFCavity))
-    u0 = get_energy_loss(ring, method=method) / ring.periodicity
+        cavpts = ring.get_cells(checktype(RFCavity))
+    u0 = ring.get_energy_loss(method=method) / ring.periodicity
     freq = numpy.array([cav.Frequency for cav in ring.select(cavpts)])
     rfv = numpy.array([cav.Voltage for cav in ring.select(cavpts)])
     tl0 = numpy.array([cav.TimeLag for cav in ring.select(cavpts)])
+    if u0 > numpy.sum(rfv):
+        raise AtError('Not enough RF voltage: unstable ring')
```

Comment:
This test is better located where it was before (in the "Else" clause): summing the voltages only makes sense if all frequencies and timelags are identical. And the "missing Voltage" error is already handled in the "Except" clause.

Reply:
Done